File: src/obsidian_mcp/folders.py
"""Folder management tools for the Obsidian vault."""

from __future__ import annotations

import os
import shutil

from mcp.server.fastmcp import FastMCP, Context

from obsidian_mcp.errors import FolderNotFoundError, FolderNotEmptyError
from obsidian_mcp.vault import Vault


def create_folder(vault: Vault, path: str) -> str:
    folder_path = vault.root / path
    folder_path.mkdir(parents=True, exist_ok=False)
    return str(folder_path.relative_to(vault.root))


def rename_folder(vault: Vault, old_path: str, new_path: str) -> str:
    src = vault.root / old_path
    if not src.exists():
        raise FolderNotFoundError(f"Folder not found: {old_path!r}")
    dst = vault.root / new_path
    shutil.move(str(src), str(dst))
    return str(dst.relative_to(vault.root))


def delete_folder(vault: Vault, path: str, force: bool = False) -> str:
    folder_path = vault.root / path
    if not folder_path.exists():
        raise FolderNotFoundError(f"Folder not found: {path!r}")

    if not force:
        contents = list(folder_path.iterdir())
        if contents:
            raise FolderNotEmptyError(
                f"Folder is not empty: {path!r}. Use force=True to delete anyway."
            )
        os.rmdir(folder_path)
    else:
        shutil.rmtree(folder_path)

    return path


def register_tools(mcp: FastMCP) -> None:
    """Register folder tools with the MCP server."""
    from mcp.server.fastmcp.exceptions import ToolError

    @mcp.tool()
    async def folder_create(path: str, ctx: Context = None) -> str:
        """Create a new folder in the vault."""
        vault: Vault = ctx.request_context.lifespan_context.vault
        try:
            return create_folder(vault, path)
        except FileExistsError as e:
            raise ToolError(str(e))
        except OSError as e:
            raise ToolError(f"I/O error: {e}")

    @mcp.tool()
    async def folder_rename(old_path: str, new_path: str, ctx: Context = None) -> str:
        """Rename or move a folder."""
        vault: Vault = ctx.request_context.lifespan_context.vault
        try:
            return rename_folder(vault, old_path, new_path)
        except FolderNotFoundError as e:
            raise ToolError(str(e))
        except OSError as e:
            raise ToolError(f"I/O error: {e}")

    @mcp.tool()
    async def folder_delete(path: str, force: bool = False, ctx: Context = None) -> str:
        """Delete a folder. Use force=True for non-empty folders."""
        vault: Vault = ctx.request_context.lifespan_context.vault
        try:
            return delete_folder(vault, path, force)
        except (FolderNotFoundError, FolderNotEmptyError) as e:
            raise ToolError(str(e))
        except OSError as e:
            raise ToolError(f"I/O error: {e}")

File: rust/src/folders.rs
use std::fs;

use crate::errors::{ObsidianError, Result};
use crate::vault::Vault;

/// Create a new folder inside the vault. Returns its relative path.
pub fn create_folder(vault: &Vault, path: &str) -> Result<String> {
    let folder_path = vault.root.join(path);
    // mkdir without exist_ok — error if already exists
    fs::create_dir_all(&folder_path)?;
    Ok(vault.relative(&folder_path))
}

/// Rename (move) a folder. Returns the new relative path.
pub fn rename_folder(vault: &Vault, old_path: &str, new_path: &str) -> Result<String> {
    let src = vault.root.join(old_path);
    if !src.exists() {
        return Err(ObsidianError::FolderNotFound(format!(
            "Folder not found: {:?}",
            old_path
        )));
    }
    let dst = vault.root.join(new_path);
    if let Some(parent) = dst.parent() {
        fs::create_dir_all(parent)?;
    }
    fs::rename(&src, &dst)?;
    Ok(vault.relative(&dst))
}

/// Delete a folder. If `force` is false, the folder must be empty.
/// Returns the deleted folder's path.
pub fn delete_folder(vault: &Vault, path: &str, force: bool) -> Result<String> {
    let folder_path = vault.root.join(path);
    if !folder_path.exists() {
        return Err(ObsidianError::FolderNotFound(format!(
            "Folder not found: {:?}",
            path
        )));
    }

    if force {
        fs::remove_dir_all(&folder_path)?;
    } else {
        // Check if empty
        let mut entries = fs::read_dir(&folder_path)?;
        if entries.next().is_some() {
            return Err(ObsidianError::FolderNotEmpty(format!(
                "Folder is not empty: {:?}. Use force=true to delete anyway.",
                path
            )));
        }
        fs::remove_dir(&folder_path)?;
    }

    Ok(path.to_string())
}
